### ISSUE TITLE
Angular 1.5.8 Ban

### ISSUE THREAD

**Issue by kspearrin** [mozilla/addons-linter#1000]
### Describe the problem and steps to reproduce it:

Submit Web Extension API extension using Angular 1.5.8
### What happened?

Linter throws banned error with linked article.
### What did you expect to happen?

Better explanation.
### Anything else we should know?

Hey guys, we've been developing our new extension (https://github.com/bitwarden/browser) for months now and are hoping to bring it to Firefox. It has been on Chrome for a month or so now and we have great demand for Firefox. After a 1 month long wait for a review by the addons team here we were asked to make a slight change to remove some minified libraries. Now we go to re-submit our extension and we see that we can no longer publish it due to a ban on Angular 1.x outright. Obviously this sucks for us since we've invested so much time developing it over the past few months and can't just simply remove Angular without starting all over.

The linked slideshare about why it has been banned (http://www.slideshare.net/x00mario/an-abusive-relationship-with-angularjs) is from December 2015 and gives no indication that this is still a vulnerability. It even says Google has resolved it.

We are using the latest versions on Angular (1.5.8) and had no plans to update to 2.x since it's still so fresh and Google claims it will continue to support the 1.x versions for a long time.

Can you please shed further light as to why 1.5.8 is now banned? Do we have any other options rather than re-write our entire extension all over again? Can a more detailed, cooperative review allow us to bypass this all out ban?

Thank you.

**Comment by tofumatt** [comment]
Paging @wagnerand, if Google is still supporting this version and it's fixed this seems like we're being too aggressive.

**Comment by wagnerand** [comment]
My information is that Google stopped supporting angular 1.x months ago. It is now a community driven project. Also, we have been made aware of exploits that are not part of the presentation.

**Comment by kspearrin** [comment]
@tofumatt @wagnerand I have never read anything from the team that says they are no longer supporting 1.x. Here are some posts that I found that have quotes from the Google team regarding support:

http://stackoverflow.com/questions/37037251/angularjs-1-x-support-lifecycle
https://www.quora.com/When-Angular-1-X-will-become-deprecated

Looking at the GitHub repo shows that the top 3 contributors over the past month are all from the Angular team themselves still:

https://github.com/angular/angular/pulse/monthly

Angular 2 was only released 1 month ago so already dropping support for Angular 1.x already seems too aggressive to me.

**Comment by tofumatt** [comment]
I agree with @kspearrin.

I know @wagnerand has mentioned other exploits (seemingly) not public; have you disclosed them to the Angular team so they can fix them? Seems like this is a bit too aggressive.

**Comment by wagnerand** [comment]
Unfortunately, we were not able to report them to angular as the security researcher who found them asked us to not share them.
The underlying issue is a design specific issue though. Even if public exploits can be fixed, there will most probably be always another way to circumvent the patches.

**Comment by tofumatt** [comment]
Ouch, that sounds brutal that they wouldn't share the security exploits with the team 😢

Okay, we should update [the docs](https://github.com/mozilla/addons-linter/blob/master/docs/third-party-libraries.md#angularjs-1x) with this extra info then; even if we can't disclose the vulnerability we can say that we know of one and don't feel comfortable with people running angular in webextensions.

**Comment by kspearrin** [comment]
Is there any possible way for us to get around this ban? Are all parts of Angular affected? Organizations with large extension projects cannot react to new versions so quickly (especially the large overhaul that Angular 2 is). It would seem to me that you are going to block out a large number of developers from distributing their extensions on Firefox since using Angular 1.x is still encouraged by Chrome in their web extension documentation today.

**Comment by joepie91** [comment]
It's *very* disturbing that the vulnerabilities are not being reported to the Angular team. That's an important part of responsible disclosure...

**Comment by justjanne** [comment]
@joepie91 But might be a legal issue, if the researcher only released the information to the FF team under a strict non-disclosure agreement.

There’s nothing they can do about it in such a case (but the researcher seems shady for not reporting it to the source indeed)

**Comment by ihsw** [comment]
@joepie91 The only responsible disclosure is full, public, and anonymous disclosure.

**Comment by ian-a-hickey** [comment]
Angular is removing the expression sandbox as of 1.6. They are doing that because the sandbox (while never intended to be used for security) was being used for security. Researchers kept showing over and over that the sandbox was not secure enough. Eventually, they just removed the sandbox. Personally, I feel like the sandbox (while not perfect) was actually pretty effective. I actually submitted the last known sandbox escape for versions 1.5 which seemed to lead to the removal of the sandbox. I now wish I hadn't.  http://angularjs.blogspot.com/2016/09/angular-16-expression-sandbox-removal.html , https://github.com/angular/angular.js/issues/14939

**Comment by Pharylon** [comment]
* Google still supports 1.x. 
* Even if they didn't, does Mozilla suddenly have a beef with community-driven projects???
* "There's a super-secret vulnerability in one the most popular web frameworks ever, seriously, I can't tell you what it is but it's totally there" doesn't pass the smell test.

**Comment by mprobst** [comment]
Hi. I'm working on security for Angular 1. We've reached out to Mozilla and the security researcher on this and are trying to clarify what's going on. We believe this is likely a misunderstanding. Please stay tuned, we'll comment on this issue.

**Comment by joepie91** [comment]
@justjanne *If* that happened (and that's highly speculative at this point), that seems like the moment for Mozilla to take a stand and say "no, we will not sign an NDA on this". NDAs for vulnerability disclosure (other than embargoes from the vendor *themselves*) are highly unusual, and for very good reason...

@ihsw This is probably not the right venue for a lengthy discussion about this, but the idea of "responsible disclosure" is to get things fixed as quickly as possible with as little impact as possible. While full disclosure is *sometimes* the answer to that (particularly with troublesome vendors), it would be wrong to equate "full disclosure" and "responsible disclosure" as if they are one and the same thing.

**Comment by AgentME** [comment]
My understanding is that Angular runs eval-like functions on the page DOM. When Angular is run from within an extension using a DOM controlled by the webpage, then the webpage can write code into its own DOM which Angular from within the extension will execute with its extension privileges. Making a general secure sandbox for this type of case is very difficult (see how complicated Google Caja is; is the Angular team going to go through that much effort purely to make Angular work in extensions?). They've probably put band-aids around a few specific escapes, but as long as eval is used on arbitrary text from the page DOM controlled by an attacker, the root cause isn't fixed and there's probably many escapes more possible.

**Comment by e-oz** [comment]
@wagnerand

>  It is now a community driven project. 

Please clarify one thing especially: all lins and frameworks, driven by community, are enough reason for ban, right?

**Comment by dosaygo-coder-0** [comment]
You have a responsibility to the people to share the vulnerability with the Angular team. Following the prohibition of the security researcher, right or wrong, at the expense of sharing information that could protect people and companies from these vulnerabilities, really doesn't seem right at all. Ultimately, it's incorrect to justify it by saying "researcher told me to do it", you have to take responsibility for your choice not to share it. So I don't think it works to simply deflect the matter to it having being already decided by someone else. There are likely issues bearing on the sharing of this information, that you could contributed to discussing in order to work out the best way forward. To silence that discussion from happening, because someone told you not to -- doesn't seem to be in everyone's benefit here.

**Comment by mooman219** [comment]
[Casual note this was posted on [HackerNews](https://news.ycombinator.com/item?id=12752248), therefore the comment influx]

**Comment by justjanne** [comment]
@dosaygo-coder-0 Many NDAs even prohibit discussing that you are under an NDA in the first place. There has been nothing confirmed yet regarding why Mozilla says they can’t discuss the issue or tell anyone about it yet, so we shouldn’t jump to conclusions just yet.

**Comment by AgentME** [comment]
@dosaygo-coder-0 

>You have a responsibility to the people to share the vulnerability with the Angular team.

The Angular team already decided that the issues of Angular within extensions were not in scope of the project: [the sandbox was specifically described as not a defense mechanism and was entirely removed in 1.6](http://angularjs.blogspot.com/2016/09/angular-16-expression-sandbox-removal.html).

**Comment by callahad** [comment]
[Thanks @mooman219 - DevRel is aware and will pay attention to the HN thread. Also emailing some folks right now and flagging this for followup next week once I get back from a conference.]

**Comment by mordocai** [comment]
@mooman219 Your hackernews link goes to the nintendo switch discussion not the firefox one

**Comment by joshmanders** [comment]
[Correct Hacker News discussion](https://news.ycombinator.com/item?id=12754199)

**Comment by whitef0x0** [comment]
This is ridiculous. Even if there is a large security vulnerability, the fact that a dev finds this out the hard way is counterproductive.

Mozilla should support Angular by disclosing vulnerabilities to the Angular team ASAP

**Comment by dosaygo-coder-0** [comment]
@justjanne indeed we shouldn't jump to conclusions. No one knows if an NDA is the excuse here or not. So far the excuse is "researcher said so." So no need to assume NDA for now. 

@AgentME assuming Angular team *might* not need to know it doesn't seem like a strong reason to prevent you from sharing it. Maybe you don't understand the extension scope view of that team, maybe you don't understand the vulnerabilities. So how can you say they don't need it for sure? And if there's a small chance, you must share it. 

Way safer to discuss everything with the relevant people, and share it with them, and decide together, than assuming you know what's best before doing that and everything's clear. Just a general point.

With @whitef0x0 agree. I think the important point here is that in the absence of total clarity, one shouldn't assume there's no need to share them. Keeping Angular team in the dark isn't going to help anyone get secured. It's reasonable to assume the vulnerabilities would be useful to them, and they are the ones who can fix it, so it must be shared, and there doesn't seem a good reason to not share them. "NDA" or "researcher" are just excuses that limit discussion and deflect responsibility, delaying the actual result of a fix from happening.

**Comment by AgentME** [comment]
@dosaygo-coder-0 The vulnerability was within a feature specifically described as not a security feature and was already removed in 1.6.

Angular runs eval on text in the page DOM. When the DOM is controlled by someone with lower privileges (ie. Angular is running in a higher-privileged extension with the DOM controlled by a webpage), then they can elevate their privileges by placing code in the DOM and letting Angular execute it within the extension.

Angular <1.6 blacklisted specific ways to do this, but it was not nearly complete. Fully securing this would take a ton of effort and would heavily bloat the size of Angular. The vulnerability mentioned in https://github.com/mozilla/addons-linter/issues/1000#issuecomment-255089138 was most likely another way around this blacklist. The blacklist was not guaranteed or even intended to be foolproof: it's described [here](http://angularjs.blogspot.com/2016/09/angular-16-expression-sandbox-removal.html) as "not a defense mechanism". The blacklist was entirely removed in 1.6. They admit that the sandbox blacklist was never enough to properly secure Angular in contexts such as extensions where an attacker can insert templates; from that post, it's apparent they're not interested in knowing more ways it could have been bypassed.

**Comment by dveditz** [comment]
I'm on the incoming end of most security bug submissions to Mozilla. We don't sign NDAs to get bug info. Besides, that would involve the lawyers and takes forever. We do sometimes agree to embargo a bug, especially if it involves multiple products (not withholding _from_ those products, but so they have time to patch also). We definitely would not agree to withhold a vulnerability from the maintainers of an upstream library if that's where it needs to be patched.

**Comment by nikoTM** [comment]
@dveditz so how come this was not forwarded to angular team ?

**Comment by thejh** [comment]
@ian-a-hickey FYI, I have a working bypass, works even in CSP mode, from the simple Angular binding context (`{{...}}`), and it still worked with your bug fixed (before they removed the whole sandbox).

**Comment by mryellow** [comment]
> You have a responsibility to the people to share the vulnerability with the Angular team.

There is no fix for broken architecture. Only fix is to stop using it for purposes it is not fit for.

**Comment by whitef0x0** [comment]
@mprobst any updates from the Angular team on this?

**Comment by mprobst** [comment]
@whitef0x0 yes.

* We understand the underlying concern better now after talking to the researcher
* The issue is unrelated to sandbox bypasses
* There is a (not super high priority) security issue that deserves fixing. It's not really specific to Angular, but Angular makes it easier to exploit.
* We believe the problem can be mitigated between Angular and Firefox. I'm working on a fix.

**Comment by hhsadiq** [comment]
This ban looks more like a personal grudge towards angular than anything else. The reasons provided by @wagnerand are not very compelling. Or probably firefox team is not very happy about the fact that google chrome has taken over its browser market share, so started to hate everything related to google.

**Comment by EliCDavis** [comment]
@hhsadiq save that for the reddit comments.

**Comment by dosaygo-coder-0** [comment]
@mryellow you don't know that's the problem here. And there is a fix for what you say anyway. Best to just share it with the team and let them decide whether there's any fix or not here. 

This really isn't an ideological issue, it's very clear what the thing that works for the community is. Even if you're going to ban Ng, give them a chance to address your concerns and improve. To choose not to give them that chance, and ban anyway, just hurts everyone.

Mr Cowan, I'm not sure we can assume as much as you are about the vulnerability. Even if it was so, it's not our place to just decide to withhold. Surely you should at least check with them first. I don't see any reason in what you said why sharing it with team was not what works.

Now Moz is saying they wouldn't do this. Seems like just a mixup here somewhere.

@garnik I agree.

**Comment by mi-g** [comment]
Since only a few people here understand what the security issue actually is, could they comment on this possible workaround: inserting as first line of the angular.min.js file being used in the Firefox extension:

`
(typeof window!=="undefined" && window.location && window.location.href && window.location.href.startsWith("resource://my-extension/")) || (function() { throw "Library loading not allowed" })();
`

This ensures angular is not injected into a Web site that could potentially exploit the vulnerability. I verified it was preventing loading angular when injected using a Firefox frame-script (`nsIFrameScriptLoader.loadFrameScript`) and add-on `sdk/page-mod` or `sdk/content/worker` modules. It also blocks the script if the site does something like `<script src="resource://.../angular.min.js"></script>`.

But this modified version of the angular library can still be used within the extension local environment, i.e in add-on panels, from code belonging to the add-on and supposedly safe.

Would that (temporary) workaround make using angular 1.x safe for use in a Firefox add-on ?

**Comment by AgentME** [comment]
Ohhh I think there's been two possible issues brought up here. The issue I've been bringing up above only applies if you use Angular within an extension content script (this does *not* describe [bitwarden](https://github.com/bitwarden/browser/blob/master/src/manifest.json)). Judging from @mi-g the other issue is... I guess I shouldn't be the first to go into it but I see how what @mprobst [said](https://github.com/mozilla/addons-linter/issues/1000#issuecomment-255269186) applies.

----

For completeness and to help everyone here be on the same page, here's a fuller description of the problem with Angular.js in content scripts. (Besides needing an extension using Angular in a content script, it requires a sandbox escape or a future version of Angular to attack right now, so I don't think I'm disclosing anything sensitive here.)

If I disable Angular 1.5.8's expression sandboxing (by applying https://github.com/angular/angular.js/commit/1547c751aa48efe7dbefef701c3df5983b04aa2e which will be included in the next release, or just commenting out the bodies of the ensure\* functions) and then make an extension which does nothing but load Angular as a content script on all web pages ([manifest.json](https://gist.github.com/AgentME/298914b107f0db45f91a833123b191c2)), and then I browse to a webpage containing this HTML:

```html
<!doctype html>
<html ng-app ng-csp>
<head><meta charset="utf-8"></head>
<body>
{{constructor.constructor('alert(Object.keys(chrome.extension))')()}}
</body>
</html>
```

Then an alert box pops up with ViewType,sendRequest,getURL,onRequest,inIncognitoContext,connect,sendMessage,onConnect,onMessage. The `chrome.extension` value is only available from within the extension's context: this means that the webpage successfully injected code into the extension which ran with the extension's content script privileges. (The webpage could have done something more interesting instead, like make a privileged cross-origin request to read your email, etc.) So if you combine this with a sandbox escape (which [multiple](https://www.youtube.com/watch?v=U4e0Remq1WQ) [people](https://github.com/mozilla/addons-linter/issues/1000#issuecomment-255219693) say they have), or the extension uses a newer *or* much older version of Angular.js, then a webpage can inject code into an the extension's context..

**Comment by gkalpak** [comment]
AFAICT, this is not specific to Angular or extensions. Allowing malicious content into your HTML code allows bad people to do bad things to anyone visiting the page (with or without Angular, with or without extensions). Furthermore, any content script that gets inject and somehow interact with a page gives more powerful tools to the malicious content that runs on the page (with or without Angular).

Do I miss something?

**Comment by AgentME** [comment]
@gkalpak I'm not sure what your point is. There's certainly no line of code in Angular that can only exist in Angular.

>Furthermore, any content script that gets inject and somehow interact with a page gives more powerful tools to the malicious content that runs on the page (with or without Angular).

That's not true of all content scripts. Most code doesn't read from the DOM and then try to eval its text. Angular does that by default.

**Comment by gkalpak** [comment]
Sorry, I wasn't clear. What I meant is that banning all content scripts that eval text from the DOM (including angular.js) may make sense. But this is different from preventing angular.js (or scripts that do similar things) from being used in popups or background pages (or generally in contexts that do not have direct access to the webpages). And afaik this is what this thread is about.

**Comment by adamchenwei** [comment]
Is this really for the security? That do not share policy does not sound right at all... What is going on with Firefox team? Did you forgot who are the ones helped your growth? Wow...

**Comment by adamchenwei** [comment]
Is it just up to 1.5.8 got banned? So 1.5.8+ is not?

**Comment by paolocaminiti** [comment]
The ban makes perfect sense, reviewers do detailed reviews, and will ban any code that can be exploited, even simple DOM manipulations via .innerHTML are forbidden. I've had several reviews failing before publishing an extension for a well trusted company and all things pointed out made sense, it's code not a question of trust and authorship. The reviewing team (volounteers) is even informative and usually discuss and points to possible solutions (as using JSONML as a view library, which btw makes things lighter you really don't even want jQuery and the likes to bloat every user's page). If Angular is not suited to run in a privileged environment, than it is the wrong tool unless it can be clearly fixed beyond doubt.

**Comment by yonjah** [comment]
I'm not sure if this is not a bit off topic (and outsie of the scope for this issue), but this security issue has two sides to it.

1. Angular evaluating and executing code out of the DOM
2. Firefox addon allowing content scripts to run inside web pages with extremely sensitive set of permissions.

Though banning angular from extension might be the immediate and temporary fix. There might be many other content script with security holes which allow to exploit this issue.

A better solution might be for content script to only have a small api allowing them to send and receive messages to a  background worker. A malicious page might still be able to exploit the content script to send commands to the background script. but he will also need to find a way to get the background script to send out and back useful information. which in most cases wouldn't be very easy (since most apps will only have a very limited api only for app needs) and would probably be a lot easier to detect and protect  (since we now have only one exploitable channel to protect)

**Comment by cure53** [comment]
FWIW, we did the original research (together with many others) and are happy to answer any technology-related questions about that very research :)

**Comment by cure53** [comment]
Hi @mrcktz,

> your team leader is the author of the slides that were linked in the rejection notice by Mozilla

That is correct, writing here in person.

> What is this fundamental design issue that differentiates angular from other libraries

AngularJS, under certain conditions, allows to bypass parts of CSP. CSP, among the many goals it has, aims to block execution of inline JavaScript (i.e. `<h1 onclick=evil()>`). AngularJS however supports `<h1 ng-click=$event.view.evil()>` when run with its own CSP-mode. That bypasses CSP *if the stars are aligned right*. The stars are known to be quite alignment-friendly here and there. 

Also please note that my example above uses an old bypass. Newer bypasses have been posted by [Google](https://sites.google.com/site/bughunteruniversity/nonvuln/angularjs-expression-sandbox-bypass) and @thejh ([here](https://twitter.com/tehjh/status/789309231585456129)). 

While I wouldn't say that the library should be _banned_ without giving it further thought, there is definitely an increase of attack surface. That is being worked on as we speak.

> Most importantly: Are there other libraries that should also be banned because of similar reasons?

Not to my knowledge, but who knows what kinds of libraries exist out there :) Probably there are.

> Why would someone report to Mozilla and only to Mozilla?

I am not aware of any conversation where that kind information was reported exclusively. 

Our (mostly poorly done) research is public [here](http://www.slideshare.net/x00mario/jsmvcomfg-to-sternly-look-at-javascript-mvc-and-templating-frameworks) and [here](http://www.slideshare.net/x00mario/an-abusive-relationship-with-angularjs) and [here](https://twitter.com/0x6D6172696F/status/757517148847738880). We had and have good contact to the AngularJS team (starting [2013](https://docs.google.com/document/d/1epha4VgFZVvauFJb2Tx_3NJlb3D91PjyZuO5YNAMX0M/edit) and lasting until today, cc @petebacondarwin @IgorMinar), Google and Mozilla. 

An exclusive, AngularJS-related report to AMO never happened to my knowledge. Would have made no sense in any way either :) Likely a misunderstanding of some kind - I first time heard of this case four days ago.

**Comment by kspearrin** [comment]
Now that all of this discussion has settled down a bit, can @wagnerand or anyone from Mozilla chime in and let me know what the plan/outlook is here? Are we waiting on changes to Angular to get patched and will reconsider the ban then or what? Is there any path around the vulnerability that can be taken in order to subvert the ban and get our extension listed in the meantime?

**Comment by mi-g** [comment]
I would add: will Mozilla reconsider the ban for extensions that do not inject Angular into external web sites where the DOM may be maliciously trapped, but keep it for local UI usage within the extension HTML code (basically serving the page from a resource:// URL) ?

**Comment by petebacondarwin** [comment]
FYI Angular and Mozilla have met to discuss this problem and are working together to find a way to mitigate any vulnerabilities, which will allow Angular 1 to be removed from the blacklist so that it can continue to be used to develop Mozilla add-ons. Not sure what the timescale is for update but I would watch this space over the next couple of weeks.

**Comment by deadsquid** [comment]
I'm Kev Needham, a Product Manager for Add-ons. We [met with the Angular team](https://twitter.com/IgorMinar/status/790925949273567233), and we’re working our way through how we got here, and how we move forward. 

There's two main pieces to this thread, and I'd like to comment (briefly) on them.

The first issue covers responsible disclosure. Earlier this year, our team acted on a vulnerability we learned of without directly involving the Mozilla Security team. We should have consulted our security team on the vulnerability and possible actions. Had we done so, the Mozilla Security team would made the Angular team aware and given them an opportunity to respond. This didn't happen.

Our intention was to act immediately to minimize the risk to end-users; but part of mitigating the risk is ensuring that the issue is corrected upstream. We made a mistake by not involving all the right people, and we're updating our internal processes to ensure this doesn't happen again. Our apologies to the Angular team for not ensuring we communicated our intents and rationale.

The second issue covers what's next with extensions that incorporate Angular. As mentioned above, we met this week with the Angular team to go over the concerns we have, and are working with them on a mitigation strategy. As @petebacondarwin mentions, it's a work in progress, and we don't have a definitive answer today. We'll be posting an update when we have a go-forward plan, but I wanted to ensure people are aware that both teams are working on this matter.

**Comment by petebacondarwin** [comment]
(And I just want to thank all of those people from both Mozilla, Google and external teams that have acted to effectively and cooperatively to get to the bottom of this so quickly once the issue came to light.)

**Comment by pankajkhanzode** [comment]
@deadsquid and @petebacondarwin - Is there any update on this? Is there mitigation path where in we can try that and mention to AMO reviewers as they have disabled/revoked the published extension now and it is hurting us badly. The other extension is stuck in review and this very same question is being asked. Thank you for the response.

**Comment by petebacondarwin** [comment]
We have patched Angular as agreed in our meetings. We are now waiting for Mozilla to unblacklist the framework from their linter.

**Comment by pankajkhanzode** [comment]
@petebacondarwin - is https://github.com/angular/angular.js/releases/tag/v1.6.0-rc.0 the release that has fixes? It is in RC - is there any released version for Angular that you could point me to? Thanks!

**Comment by petebacondarwin** [comment]
We plan to have 1.5.9 and 1.6.0.rc.1 out this week that include the patch. 1.6.0 should be released in the next week or two.

**Comment by pankajkhanzode** [comment]
Thanks @petebacondarwin - I will watch out for the update on 1.5.9

**Comment by rabbitfang** [comment]
It looks like this is the patch that was implemented: https://github.com/angular/angular.js/commit/0ff10e1b56c6b7c4ac465e35c96a5886e294bac5

**Comment by mprobst** [comment]
@rabbitfang yes, that's the patch.

It's maybe worth noting that this is a defense in depth issue. An exploit would have required a website that already has an XSS bug but is using CSP to protect against itself (which is very rare), and a site visitor that uses Firefox and has a predictable addon installed that contains Angular at a known location. At that, it's still worth fixing, but highly unlikely that this is exploited in the wild on a substantial scale.

There is no security issue with using Angular in Firefox addons per se, nor with using Angular on arbitrary websites (at least none that we're aware of).

Something for Firefox addon developers to be aware of though: make sure you do not inject Angular into arbitrary websites, that's a very bad security practice and certain to cause all kinds of problems.

**Comment by pankajkhanzode** [comment]
@petebacondarwin - Could you please provide update about 1.5.9 release that is getting out official? I didn't see under releases. I would like to take dependency on released one, instead of checkin. Also, will you loop back with Kev Needham, a Product Manager for Add-ons (@deadsquid) to relook at Angular 1.5.9 and unblacklist it from being used, so that we can go ahead with our plans?

**Comment by Rob--W** [comment]
@petebacondarwin @mprobst That check for `resource:` and `chrome-extension:` is insufficient.

You must also include: `moz-extension:` (=Firefox WebExtensions), `ms-browser-extension:` (=Edge extension), `safari-extension:` (=Safari extensions).

There are probably other extension environments that I did not mention. How about disallowing bootstrapping if the protocols do not match and the requested resource is not a default scheme for the web? I.e. allowed iff the scheme is one of: http, https, ftp (?), blob, filesystem, data, file

**Comment by petebacondarwin** [comment]
Thanks for the review @Rob--W - I will talk with @mprobst in the USA morning and come back with a response.

**Comment by pankajkhanzode** [comment]
@petebacondarwin - Any update on progress made here? We have been waiting for weeks now, and while long term we plan to move to Angular 2.0, it is definitely not couple of days work to move there. So we are depending on your interaction and making Angular 1.5.* go ahead and get white-listed again. Thank you very much!

**Comment by mi-g** [comment]
Waiting here for the final call too, with a few million users depending on that decision.

**Comment by petebacondarwin** [comment]
I will have an update on changes to our patch if necessary by the close of US daytime today.

**Comment by IgorMinar** [comment]
Hello everyone, on behalf of the Angular team I'm posting an update after our meeting with the Mozilla folks today.

These were the key take aways from the meeting:

- Mozilla folks agreed that to the change angular/angular.js#15346 we implemented 3 weeks ago is a good workaround for the problem identified as "Firefox script loader can be tricked into loading angular.js into websites from Firefox addons, where once angular autobootstraps bad things could happen"
  - Mozilla security folks requested a few minor tweaks to the implementation that we already implemented as angular/angular.js#15427 right after our meeting
- Mozilla folks expressed concerns about the use case where addons inject Angular into web pages via a content script. The number of addons that do this is currently limited.
  - Angular team believes that this is an anti-pattern due to concerns about the impact of loading Angular into a foreign environment that could cause various unexpected collisions. We recommended that such addons are banned until the day when this kind of usage is officially supported by Angular - this is however currently not on our roadmap.
- No other significant concerns were brought to our attention by the Mozilla team.
  - Angular team committed to release version 1.5.9 that will contain the two pull requests mentioned above tomorrow (Nov 24).
  - Mozilla team committed to lifting the ban for all version of Angular >=1.5.9 effective Dec 1, 2016 when they'll push an update to their addon linting infrastructure.

I'm hopeful that after weeks of delays, this issue will be finally resolved on December 1, 2016. In the meantime, happy Thanksgiving to those observing Thanksgiving tomorrow! 🦃

@deadsquid please confirm this agreement as the Mozilla representative. If I accidentally got anything wrong, please point it out. Thank you!

**Comment by pankajkhanzode** [comment]
Thanks a lot @IgorMinar and Angular team for resolving it with Mozilla team. Happy ThankssGiving!

**Comment by SebastianMetzger** [comment]
@IgorMinar This is very good news, I already made a calendar entry for a december 1st release for our team 💯 
@deadsquid Please confirm 👍

**Comment by petebacondarwin** [comment]
We have just released 1.5.9 containing the security patches requested by Mozilla. It is available right now on npm and code.angularjs.org; it will be available on the Google CDN tomorrow.

We are also releasing 1.6.0-rc.2 containing the same patches later today.

We expect that Mozilla will lift its ban on add-ons built with Angular 1.5.9+ by December 1st.

**Comment by pankajkhanzode** [comment]
Thanks @petebacondarwin , @deadsquid, @IgorMinar and @mprobst + @Rob--W - for working together. 1.5.9 is great news!

**Comment by deadsquid** [comment]
Howdy folks, 

We should have a review policy update out early next week and a push to the linter by end-of-day (Pacific) time on Dec 1st, per @IgorMinar's post above. 

It's important to note a couple things:

- If your add-on uses Angular, you will need to include unmodified, non-repackaged versions. Add-ons that incorporate modified versions will be rejected.
- We'll be reaching out to affected add-on developers to update their extension to incorporate the patched version of Angular. Please ensure you update your add-ons as soon as possible with a version of Angular => 1.5.9.

Thanks again to the Angular team for working through the issues discovered with us, and for incorporating the additional changes.

**Comment by andymckay** [comment]
Just a note that this is scheduled to be deployed Dec 1st as per the AMO deployment schedule calendar: https://calendar.google.com/calendar/embed?src=mozilla.com_lr5jsh38i6dmr72uu4d1nv7dcc%40group.calendar.google.com

**Comment by cure53** [comment]
Bypassed the fix, wrote @petebacondarwin :)

**Comment by petebacondarwin** [comment]
Aw man!

**Comment by petebacondarwin** [comment]
@cure53 - we have landed fixes for the bypasses in master, 1.6 and 1.5 branches. See:

* https://github.com/angular/angular.js/commit/ebe90051eda8a3328e5993cca1663e28d03113d0
* https://github.com/angular/angular.js/commit/a649758655843275cc477fb638f8e55f72a4eaa6

**Comment by kspearrin** [comment]
@petebacondarwin  What 1.5.x version can we target for this fix?

**Comment by cure53** [comment]
@petebacondarwin LGTM, @masatokinugawa what do you think?

**Comment by petebacondarwin** [comment]
@kspearrin - we were hoping not to do any more 1.5.x releases, but I guess we might have to make an exception in this case. The difficulty is that we must get 1.5.x commits synced into Google before we can release so it will take a little longer. The version number will be 1.5.12

**Comment by kspearrin** [comment]
@petebacondarwin If 1.6.x is the only upgrade path that is fine, we'll just have to make that move sooner than originally planned. I was just under the impression you would be releasing 1.5.x as well since you mentioned the change was applied in that branch as well.

**Comment by masatokinugawa** [comment]
I found it still can be bypassed on a specific situation.
@petebacondarwin Shall I report details to pete@ba[...]com?

**Comment by andymckay** [comment]
This issue is closed and emails quite a lot of people in the mozilla add-ons team. If you have other issues could you please report them to the angular team using their reporting system. Thanks.
